File: nnstudio/plugin-api/src/PluginManifest.cpp
/* ============================================================================
 * PluginManifest.cpp  —  Plugin manifest parsing and validation
 * LGPL v3
 *
 * Minimal JSON field extractor (no external dependency).
 * Handles the small, well-defined plugin/TUP manifest schemas.
 * Replace with nlohmann::json when the project adds it as a dependency.
 * ============================================================================ */

#include <plugin-api/include/PluginManifest.h>

#include <algorithm>
#include <cctype>
#include <fstream>
#include <sstream>

namespace nnstudio::plugin_api {

using nnstudio::core::Error;
using nnstudio::core::Result;

/* ─── Tiny JSON scanner helpers ──────────────────────────────────────────── */

namespace {

/** Return true if `text` contains `"key"` somewhere. */
static bool hasKey(const std::string& text, const std::string& key) noexcept {
    std::string needle = "\"" + key + "\"";
    return text.find(needle) != std::string::npos;
}

/**
 * Extract the first string value for `"key": "..."` in text.
 * Returns empty string if not found.
 */
static std::string extractString(const std::string& text,
                                  const std::string& key) noexcept {
    std::string needle = "\"" + key + "\"";
    auto pos = text.find(needle);
    if (pos == std::string::npos) return {};

    // skip past key + optional whitespace + colon + optional whitespace
    pos += needle.size();
    while (pos < text.size() && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\n' || text[pos] == '\r'))
        ++pos;
    if (pos >= text.size() || text[pos] != ':') return {};
    ++pos;
    while (pos < text.size() && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\n' || text[pos] == '\r'))
        ++pos;
    if (pos >= text.size() || text[pos] != '"') return {};
    ++pos;  // skip opening quote

    std::string result;
    while (pos < text.size() && text[pos] != '"') {
        if (text[pos] == '\\' && pos + 1 < text.size()) {
            ++pos;  // skip escape prefix; simplified — just copy escaped char
            result += text[pos];
        } else {
            result += text[pos];
        }
        ++pos;
    }
    return result;
}

/**
 * Extract the first integer value for `"key": <digits>` in text.
 * Returns -1 if not found or not an integer.
 */
static int extractInt(const std::string& text, const std::string& key) noexcept {
    std::string needle = "\"" + key + "\"";
    auto pos = text.find(needle);
    if (pos == std::string::npos) return -1;

    pos += needle.size();
    while (pos < text.size() && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\n' || text[pos] == '\r'))
        ++pos;
    if (pos >= text.size() || text[pos] != ':') return -1;
    ++pos;
    while (pos < text.size() && (text[pos] == ' ' || text[pos] == '\t' || text[pos] == '\n' || text[pos] == '\r'))
        ++pos;
    if (pos >= text.size() || !std::isdigit(static_cast<unsigned char>(text[pos]))) return -1;

    int val = 0;
    while (pos < text.size() && std::isdigit(static_cast<unsigned char>(text[pos])))
        val = val * 10 + (text[pos++] - '0');
    return val;
}

/**
 * Extract all string items from the first JSON array `"key": [...]` in text.
 */
static std::vector<std::string> extractStringArray(const std::string& text,
                                                    const std::string& key) noexcept {
    std::string needle = "\"" + key + "\"";
    auto pos = text.find(needle);
    if (pos == std::string::npos) return {};

    pos += needle.size();
    // find '[' after key: value separator
    while (pos < text.size() && text[pos] != '[') ++pos;
    if (pos >= text.size()) return {};
    ++pos;  // skip '['

    std::vector<std::string> items;
    while (pos < text.size() && text[pos] != ']') {
        while (pos < text.size() && text[pos] != '"' && text[pos] != ']') ++pos;
        if (pos >= text.size() || text[pos] == ']') break;
        ++pos;  // skip '"'
        std::string item;
        while (pos < text.size() && text[pos] != '"')
            item += text[pos++];
        if (!item.empty()) items.push_back(std::move(item));
        if (pos < text.size()) ++pos;  // skip closing '"'
    }
    return items;
}

/**
 * Check if a string matches the basic semver pattern x.y.z (no pre-release here).
 */
static bool isSemver(const std::string& s) noexcept {
    if (s.empty()) return false;
    int dots = 0;
    for (char c : s) {
        if (c == '.') { ++dots; continue; }
        if (!std::isdigit(static_cast<unsigned char>(c))) return false;
    }
    return dots == 2 && s.front() != '.' && s.back() != '.';
}

/**
 * Check if a string is a valid hex SHA-256 (64 lowercase hex digits).
 */
static bool isSha256(const std::string& s) noexcept {
    if (s.size() != 64) return false;
    for (char c : s)
        if (!std::isxdigit(static_cast<unsigned char>(c))) return false;
    return true;
}

/**
 * Read entire file contents into a string.
 */
static Result<std::string> readFile(const std::string& path) noexcept {
    std::ifstream ifs(path);
    if (!ifs.is_open())
        return Result<std::string>::err(Error{"ManifestLoader: cannot open file: " + path});
    std::ostringstream ss;
    ss << ifs.rdbuf();
    return Result<std::string>::ok(ss.str());
}

/**
 * Extract implementation blocks from json text for "cpp" and "python".
 * Returns a vector of PluginImplementation (0-2 entries).
 */
static std::vector<PluginImplementation> extractImplementations(
    const std::string& text) noexcept
{
    std::vector<PluginImplementation> impls;
    for (const char* lang : {"cpp", "python"}) {
        std::string needle = "\"" + std::string(lang) + "\"";
        auto pos = text.find(needle);
        if (pos == std::string::npos) continue;

        // Find the '{' opening the sub-object
        auto blockStart = text.find('{', pos + needle.size());
        if (blockStart == std::string::npos) continue;

        // Find matching '}'
        int depth = 0;
        auto blockEnd = blockStart;
        while (blockEnd < text.size()) {
            if (text[blockEnd] == '{') ++depth;
            else if (text[blockEnd] == '}') { if (--depth == 0) break; }
            ++blockEnd;
        }

        std::string block = text.substr(blockStart, blockEnd - blockStart + 1);

        PluginImplementation impl;
        impl.language = lang;
        if (std::string(lang) == "cpp") {
            impl.artifact = extractString(block, "library");
        } else {
            impl.artifact = extractString(block, "module");
            impl.wheel    = extractString(block, "wheel");
        }
        impl.sha256   = extractString(block, "sha256");

        if (!impl.artifact.empty() && !impl.sha256.empty())
            impls.push_back(std::move(impl));
    }
    return impls;
}

} // anonymous namespace

/* ─── PluginType helpers ─────────────────────────────────────────────────── */

PluginType pluginTypeFromString(const std::string& s) noexcept {
    if (s == "LAYER")          return PluginType::Layer;
    if (s == "ACTIVATION")     return PluginType::Activation;
    if (s == "OPTIMIZER")      return PluginType::Optimizer;
    if (s == "TOKENIZER")      return PluginType::Tokenizer;
    if (s == "BACKEND")        return PluginType::Backend;
    if (s == "INPUT_ADAPTER")  return PluginType::InputAdapter;
    if (s == "OUTPUT_ADAPTER") return PluginType::OutputAdapter;
    if (s == "CONTEXT_SOURCE") return PluginType::ContextSource;
    if (s == "RUNNER_CLIENT")  return PluginType::RunnerClient;
    if (s == "UI_PANEL")       return PluginType::UiPanel;
    return PluginType::Unknown;
}

/* ─── ManifestLoader::loadPlugin ─────────────────────────────────────────── */

Result<PluginManifest> ManifestLoader::loadPlugin(const std::string& json) noexcept {
    PluginManifest m;

    // manifestVersion
    m.manifestVersion = extractInt(json, "manifestVersion");
    if (m.manifestVersion != 1)
        return Result<PluginManifest>::err(Error{"manifest: 'manifestVersion' must be 1"});

    // type (reject TRUST_UPDATE)
    std::string typeStr = extractString(json, "type");
    if (typeStr == "TRUST_UPDATE")
        return Result<PluginManifest>::err(
            Error{"manifest: type 'TRUST_UPDATE' is reserved for TUP packages; use tup.manifest.json"});
    m.type = pluginTypeFromString(typeStr);
    if (m.type == PluginType::Unknown)
        return Result<PluginManifest>::err(Error{"manifest: unknown plugin type '" + typeStr + "'"});

    // required string fields
    m.id = extractString(json, "id");
    if (m.id.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'id' is required"});

    m.name = extractString(json, "name");
    if (m.name.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'name' is required"});

    m.version = extractString(json, "version");
    if (m.version.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'version' is required"});

    m.author = extractString(json, "author");
    if (m.author.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'author' is required"});

    m.license = extractString(json, "license");
    if (m.license.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'license' is required"});

    m.minStudioVersion = extractString(json, "minStudioVersion");
    if (m.minStudioVersion.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'minStudioVersion' is required"});

    if (!isSemver(m.minStudioVersion))
        return Result<PluginManifest>::err(
            Error{"manifest: 'minStudioVersion' is not a valid x.y.z semver: " + m.minStudioVersion});

    m.signature = extractString(json, "signature");
    if (m.signature.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'signature' is required"});

    // optional string fields
    m.maxStudioVersion = extractString(json, "maxStudioVersion");
    m.description      = extractString(json, "description");
    m.kbRef            = extractString(json, "kbRef");

    // capabilities (required, non-empty array)
    m.capabilities = extractStringArray(json, "capabilities");
    if (m.capabilities.empty())
        return Result<PluginManifest>::err(Error{"manifest: 'capabilities' must have at least one entry"});

    // implementations (at least one)
    m.implementations = extractImplementations(json);
    if (m.implementations.empty())
        return Result<PluginManifest>::err(
            Error{"manifest: 'implementations' must have at least one of 'cpp' or 'python'"});

    // sha256 validation
    for (const auto& impl : m.implementations) {
        if (!isSha256(impl.sha256))
            return Result<PluginManifest>::err(
                Error{"manifest: sha256 for '" + impl.language + "' implementation is not a valid 64-char hex string"});
    }

    return Result<PluginManifest>::ok(std::move(m));
}

/* ─── ManifestLoader::loadTup ────────────────────────────────────────────── */

Result<TupManifest> ManifestLoader::loadTup(const std::string& json) noexcept {
    TupManifest t;

    t.manifestVersion = extractInt(json, "manifestVersion");
    if (t.manifestVersion != 1)
        return Result<TupManifest>::err(Error{"tup: 'manifestVersion' must be 1"});

    // type must be TRUST_UPDATE
    std::string typeStr = extractString(json, "type");
    if (typeStr != "TRUST_UPDATE")
        return Result<TupManifest>::err(Error{"tup: 'type' must be 'TRUST_UPDATE'"});

    t.id = extractString(json, "id");
    if (t.id.empty())
        return Result<TupManifest>::err(Error{"tup: 'id' is required"});

    t.version = extractString(json, "version");
    if (t.version.empty())
        return Result<TupManifest>::err(Error{"tup: 'version' is required"});

    t.issuedAt = extractString(json, "issuedAt");
    if (t.issuedAt.empty())
        return Result<TupManifest>::err(Error{"tup: 'issuedAt' is required"});

    t.issuerSubject = extractString(json, "issuerSubject");
    if (t.issuerSubject.empty())
        return Result<TupManifest>::err(Error{"tup: 'issuerSubject' is required"});

    t.signature = extractString(json, "signature");
    if (t.signature.empty())
        return Result<TupManifest>::err(Error{"tup: 'signature' is required"});

    t.description = extractString(json, "description");

    // addCerts and revokeCerts are validated presence-only here;
    // deep cert parsing happens in TrustUpdateHandler.
    t.selfRevocationGuard    = !hasKey(json, "selfRevocationGuard")  ? true
                               : (json.find("\"selfRevocationGuard\": false") != std::string::npos ? false : true);
    t.requireUserConfirmation= !hasKey(json, "requireUserConfirmation") ? true
                               : (json.find("\"requireUserConfirmation\": false") != std::string::npos ? false : true);

    return Result<TupManifest>::ok(std::move(t));
}

/* ─── File-based loaders ─────────────────────────────────────────────────── */

Result<PluginManifest> ManifestLoader::loadPluginFromFile(const std::string& path) noexcept {
    auto text = readFile(path);
    if (!text.ok()) return Result<PluginManifest>::err(text.error());
    return loadPlugin(text.value());
}

Result<TupManifest> ManifestLoader::loadTupFromFile(const std::string& path) noexcept {
    auto text = readFile(path);
    if (!text.ok()) return Result<TupManifest>::err(text.error());
    return loadTup(text.value());
}

} // namespace nnstudio::plugin_api

File: nnspire/python-bridge/nnspire/plugin_manifest.py
"""
NNSpire.plugin_manifest — Plugin and TUP manifest parsing / validation.

Validates plugin.manifest.json and tup.manifest.json files against the same
rules as the C++ ManifestLoader.  Optionally uses the ``jsonschema`` package
(``pip install jsonschema``) for full JSON Schema Draft 7 validation; falls
back to manual field checking when ``jsonschema`` is not installed.

USAGE
─────
    from NNSpire.plugin_manifest import load_plugin_manifest, load_tup_manifest

    # Validate and parse a plugin manifest
    result = load_plugin_manifest("path/to/plugin.manifest.json")
    if result.ok:
        m = result.value
        print(m.id, m.version, m.type)
    else:
        print("Invalid manifest:", result.error)

    # Validate and parse a TUP manifest
    result = load_tup_manifest("path/to/tup.manifest.json")

@kb: PLUGIN-SDK.md §4  |  TRUST-ARCHITECTURE.md §TUP
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field
from typing import List, Optional

# ─── Optional jsonschema ────────────────────────────────────────────────────

try:
    import jsonschema  # type: ignore
    _HAS_JSONSCHEMA = True
except ImportError:
    _HAS_JSONSCHEMA = False

# Path to the schema files relative to this module
_PKG_DIR    = os.path.dirname(os.path.abspath(__file__))
_SCHEMA_DIR = os.path.normpath(
    os.path.join(_PKG_DIR, "..", "plugin-api", "schemas")
)

# ─── Result type ────────────────────────────────────────────────────────────

@dataclass
class ValidationResult:
    """Simple Result monad for manifest validation."""
    ok: bool
    value: object = None
    error: str = ""

    @staticmethod
    def success(value: object) -> "ValidationResult":
        return ValidationResult(ok=True, value=value)

    @staticmethod
    def failure(message: str) -> "ValidationResult":
        return ValidationResult(ok=False, error=message)


# ─── Data classes ────────────────────────────────────────────────────────────

@dataclass
class PluginImplementation:
    language: str           # "cpp" or "python"
    artifact: str           # library path or module name
    sha256: str             # hex SHA-256
    wheel: str = ""         # Python wheel path (optional)


@dataclass
class PluginManifest:
    manifest_version: int
    id: str
    name: str
    version: str
    type: str               # e.g. "LAYER", "TOKENIZER"
    author: str
    license: str
    min_studio_version: str
    capabilities: List[str]
    implementations: List[PluginImplementation]
    signature: str
    max_studio_version: str = ""
    description: str = ""
    kb_ref: str = ""
    author_url: str = ""
    license_url: str = ""


@dataclass
class TupCertEntry:
    file: str
    sha256: str
    role: str               # "root", "intermediate", or "crl"
    subject: str = ""


@dataclass
class TupRevocation:
    serial_number: str
    issuer_subject: str
    reason: str
    comment: str = ""


@dataclass
class TupManifest:
    manifest_version: int
    id: str                 # UUID v4
    version: str
    issued_at: str          # ISO 8601
    issuer_subject: str
    add_certs: List[TupCertEntry]
    revoke_certs: List[TupRevocation]
    signature: str
    description: str = ""
    self_revocation_guard: bool = True
    require_user_confirmation: bool = True


# ─── Patterns ────────────────────────────────────────────────────────────────

_SEMVER_RE   = re.compile(r"^\d+\.\d+\.\d+$")
_SHA256_RE   = re.compile(r"^[0-9a-f]{64}$")
_UUID4_RE    = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
)
_RDOMAIN_RE  = re.compile(r"^[a-z][a-z0-9]*(\.[a-z][a-z0-9-]*)+$")

_VALID_TYPES = {
    "LAYER", "ACTIVATION", "OPTIMIZER", "TOKENIZER", "BACKEND",
    "INPUT_ADAPTER", "OUTPUT_ADAPTER", "CONTEXT_SOURCE", "RUNNER_CLIENT",
    "UI_PANEL",
}  # TRUST_UPDATE excluded — that path uses tup.manifest.json


# ─── Internal helpers ────────────────────────────────────────────────────────

def _require(d: dict, key: str, context: str = "") -> ValidationResult:
    prefix = f"{context}: " if context else ""
    if key not in d:
        return ValidationResult.failure(f"{prefix}missing required field '{key}'")
    val = d[key]
    if val is None or val == "" or val == [] or val == {}:
        return ValidationResult.failure(f"{prefix}field '{key}' must not be empty")
    return ValidationResult.success(val)


def _schema_validate(data: dict, schema_name: str) -> Optional[str]:
    """
    Run jsonschema validation if available.
    Returns an error string on failure, or None on success.
    """
    if not _HAS_JSONSCHEMA:
        return None
    schema_path = os.path.join(_SCHEMA_DIR, schema_name)
    if not os.path.exists(schema_path):
        return None
    with open(schema_path, encoding="utf-8") as fh:
        schema = json.load(fh)
    try:
        jsonschema.validate(instance=data, schema=schema)
        return None
    except jsonschema.ValidationError as exc:
        return str(exc.message)


# ─── Plugin manifest loading ─────────────────────────────────────────────────

def _parse_implementations(raw: dict) -> ValidationResult:
    impls_raw = raw.get("implementations", {})
    if not impls_raw:
        return ValidationResult.failure("'implementations' must have at least one of 'cpp' or 'python'")

    impls: List[PluginImplementation] = []
    for lang in ("cpp", "python"):
        entry = impls_raw.get(lang)
        if not entry:
            continue
        if lang == "cpp":
            artifact = entry.get("library", "")
        else:
            artifact = entry.get("module", "")
        sha256 = entry.get("sha256", "")
        wheel  = entry.get("wheel", "")
        if not artifact:
            return ValidationResult.failure(
                f"implementations.{lang}: missing artifact field (library/module)"
            )
        if not sha256 or not _SHA256_RE.match(sha256):
            return ValidationResult.failure(
                f"implementations.{lang}: sha256 must be a 64-char lowercase hex string"
            )
        impls.append(PluginImplementation(language=lang, artifact=artifact,
                                           sha256=sha256, wheel=wheel))
    if not impls:
        return ValidationResult.failure(
            "'implementations' must have at least one of 'cpp' or 'python'"
        )
    return ValidationResult.success(impls)


def load_plugin_manifest(path_or_text: str) -> ValidationResult:
    """
    Parse and validate a plugin.manifest.json.

    `path_or_text` can be either:
    - A file-system path to a .json file, or
    - A raw JSON string (must start with ``{``).

    Returns a ``ValidationResult`` whose ``value`` is a ``PluginManifest``
    on success, or whose ``error`` describes the first violation.
    """
    if path_or_text.strip().startswith("{"):
        json_text = path_or_text
    else:
        try:
            with open(path_or_text, encoding="utf-8") as fh:
                json_text = fh.read()
        except OSError as exc:
            return ValidationResult.failure(f"cannot read manifest file: {exc}")

    try:
        raw = json.loads(json_text)
    except json.JSONDecodeError as exc:
        return ValidationResult.failure(f"invalid JSON: {exc}")

    # ── JSON Schema Draft 7 validation (if jsonschema is available) ──────────
    schema_err = _schema_validate(raw, "plugin.manifest.schema.json")
    if schema_err:
        return ValidationResult.failure(schema_err)

    # ── Manual required-field checks ────────────────────────────────────────
    ctx = "plugin.manifest"

    for req in ("manifestVersion", "id", "name", "version", "type",
                "author", "license", "minStudioVersion", "capabilities",
                "implementations", "signature"):
        rv = _require(raw, req, ctx)
        if not rv.ok:
            return rv

    if raw.get("manifestVersion") != 1:
        return ValidationResult.failure(f"{ctx}: 'manifestVersion' must be 1")

    plugin_type = raw["type"]
    if plugin_type == "TRUST_UPDATE":
        return ValidationResult.failure(
            f"{ctx}: type 'TRUST_UPDATE' is reserved for TUP packages"
        )
    if plugin_type not in _VALID_TYPES:
        return ValidationResult.failure(f"{ctx}: unknown plugin type '{plugin_type}'")

    if not _RDOMAIN_RE.match(raw["id"]):
        return ValidationResult.failure(
            f"{ctx}: 'id' must be a reverse-domain identifier, e.g. com.example.my-layer"
        )

    if not _SEMVER_RE.match(raw.get("version", "")):
        return ValidationResult.failure(
            f"{ctx}: 'version' must be an x.y.z semver string"
        )

    if not _SEMVER_RE.match(raw.get("minStudioVersion", "")):
        return ValidationResult.failure(
            f"{ctx}: 'minStudioVersion' must be an x.y.z semver string"
        )

    if not isinstance(raw.get("capabilities"), list) or not raw["capabilities"]:
        return ValidationResult.failure(
            f"{ctx}: 'capabilities' must be a non-empty array"
        )

    impls_rv = _parse_implementations(raw)
    if not impls_rv.ok:
        return impls_rv

    m = PluginManifest(
        manifest_version    = raw["manifestVersion"],
        id                  = raw["id"],
        name                = raw["name"],
        version             = raw["version"],
        type                = raw["type"],
        author              = raw["author"],
        license             = raw["license"],
        min_studio_version  = raw["minStudioVersion"],
        capabilities        = list(raw["capabilities"]),
        implementations     = impls_rv.value,
        signature           = raw["signature"],
        max_studio_version  = raw.get("maxStudioVersion", ""),
        description         = raw.get("description", ""),
        kb_ref              = raw.get("kbRef", ""),
        author_url          = raw.get("authorUrl", ""),
        license_url         = raw.get("licenseUrl", ""),
    )
    return ValidationResult.success(m)


# ─── TUP manifest loading ────────────────────────────────────────────────────

def load_tup_manifest(path_or_text: str) -> ValidationResult:
    """
    Parse and validate a tup.manifest.json.

    `path_or_text` can be a file-system path or a raw JSON string.

    Returns a ``ValidationResult`` whose ``value`` is a ``TupManifest``
    on success, or whose ``error`` describes the first violation.
    """
    if path_or_text.strip().startswith("{"):
        json_text = path_or_text
    else:
        try:
            with open(path_or_text, encoding="utf-8") as fh:
                json_text = fh.read()
        except OSError as exc:
            return ValidationResult.failure(f"cannot read TUP manifest: {exc}")

    try:
        raw = json.loads(json_text)
    except json.JSONDecodeError as exc:
        return ValidationResult.failure(f"invalid JSON: {exc}")

    # ── JSON Schema Draft 7 validation ───────────────────────────────────────
    schema_err = _schema_validate(raw, "tup.manifest.schema.json")
    if schema_err:
        return ValidationResult.failure(schema_err)

    # ── Manual required-field checks ────────────────────────────────────────
    ctx = "tup.manifest"

    for req in ("manifestVersion", "id", "type", "version",
                "issuedAt", "issuerSubject", "addCerts", "revokeCerts", "signature"):
        rv = _require(raw, req, ctx)
        if not rv.ok:
            # addCerts and revokeCerts may be empty lists — only check presence
            if req in ("addCerts", "revokeCerts") and req in raw:
                continue
            return rv

    if raw.get("manifestVersion") != 1:
        return ValidationResult.failure(f"{ctx}: 'manifestVersion' must be 1")

    if raw.get("type") != "TRUST_UPDATE":
        return ValidationResult.failure(f"{ctx}: 'type' must be 'TRUST_UPDATE'")

    if not _UUID4_RE.match(raw.get("id", "")):
        return ValidationResult.failure(f"{ctx}: 'id' must be a valid UUID v4")

    if not _SEMVER_RE.match(raw.get("version", "")):
        return ValidationResult.failure(
            f"{ctx}: 'version' must be an x.y.z semver string"
        )

    add_certs: List[TupCertEntry] = []
    for entry in raw.get("addCerts", []):
        file_   = entry.get("file", "")
        sha256_ = entry.get("sha256", "")
        role_   = entry.get("role", "")
        if not file_ or not sha256_ or not role_:
            return ValidationResult.failure(
                f"{ctx}: each addCerts entry must have 'file', 'sha256', and 'role'"
            )
        if not _SHA256_RE.match(sha256_):
            return ValidationResult.failure(
                f"{ctx}: addCerts entry '{file_}': sha256 must be 64 hex chars"
            )
        if role_ not in ("root", "intermediate", "crl"):
            return ValidationResult.failure(
                f"{ctx}: addCerts entry '{file_}': role must be 'root', 'intermediate', or 'crl'"
            )
        add_certs.append(TupCertEntry(
            file=file_, sha256=sha256_, role=role_,
            subject=entry.get("subject", "")
        ))

    revoke_certs: List[TupRevocation] = []
    valid_reasons = {
        "keyCompromise", "caCompromise", "affiliationChanged",
        "superseded", "cessationOfOperation", "unspecified"
    }
    for entry in raw.get("revokeCerts", []):
        sn     = entry.get("serialNumber", "")
        issuer = entry.get("issuerSubject", "")
        reason = entry.get("reason", "")
        if not sn or not issuer or not reason:
            return ValidationResult.failure(
                f"{ctx}: each revokeCerts entry must have 'serialNumber', 'issuerSubject', and 'reason'"
            )
        if reason not in valid_reasons:
            return ValidationResult.failure(
                f"{ctx}: revokeCerts entry: unknown reason '{reason}'"
            )
        revoke_certs.append(TupRevocation(
            serial_number=sn, issuer_subject=issuer,
            reason=reason, comment=entry.get("comment", "")
        ))

    t = TupManifest(
        manifest_version          = raw["manifestVersion"],
        id                        = raw["id"],
        version                   = raw["version"],
        issued_at                 = raw["issuedAt"],
        issuer_subject            = raw["issuerSubject"],
        add_certs                 = add_certs,
        revoke_certs              = revoke_certs,
        signature                 = raw["signature"],
        description               = raw.get("description", ""),
        self_revocation_guard     = raw.get("selfRevocationGuard", True),
        require_user_confirmation = raw.get("requireUserConfirmation", True),
    )
    return ValidationResult.success(t)
